scripts/copy_to_distributed_table.sh

```diff
@@ -0,0 +1,95 @@
+#!/bin/sh
+
+format='text'
+options='OIDS false, FREEZE false'
+schema='public'
+
+while getopts ':BCc:d:e:Hh:n:q:T' o; do
+	case "${o}" in
+		B)
+			format='binary'
+			;;
+		C)
+			format='csv'
+			;;
+		c)
+			encoding=`echo ${OPTARG} | sed s/\'/\'\'/g`
+			options="${options}, ENCODING '${encoding}'"
+			;;
+		d)
+			delimiter=`echo ${OPTARG} | sed s/\'/\'\'/g`
+			options="${options}, DELIMITER '${delimiter}'"
+			;;
+		e)
+			escape=`echo ${OPTARG} | sed s/\'/\'\'/g`
+			options="${options}, ESCAPE '${escape}'"
+			;;
+		H)
+			options="${options}, HEADER true"
+			;;
+		h)
+			usage
+			;;
+		n)
+			null=`echo ${OPTARG} | sed s/\'/\'\'/g`
+			options="${options}, NULL '${null}'"
+			;;
+		q)
+			quote=`echo ${OPTARG} | sed s/\'/\'\'/g`
+			options="${options}, QUOTE '${quote}'"
+			;;
+		s)
+			schema=`echo ${OPTARG} | sed s/\'/\'\'/g`
+			;;
+		T)
+			format='text'
+			;;
+        *)
+            usage
+            ;;
+    esac
+done
+shift $((OPTIND-1))
+
+options="${options}, FORMAT ${format}"
+filename=$1
+tablename=$2
+
+if [ -z "${filename}" ] || [ -z "${tablename}" ]; then
+	usage
+fi
+
+filename=`echo ${filename} | sed s/\'/\'\'/g`
+facadename=`echo "${tablename}_copy_facade" | sed s/\"/\"\"/g`
```

Comment:
Anything `TEMPORARY` in PostgreSQL is session-bound and session-isolated, i.e. it will disappear at the end of the session and is not visible _at all_ to other sessions. So it is safe to run the script:

  * Multiple times in sequence
  * Simultaneously in multiple processes

The old script used a lock because I didn't know you could make temporary functions. `CREATE TABLE` and `CREATE SEQUENCE` have `TEMPORARY` as part of their syntax, but `CREATE FUNCTION` does not, so I thought that meant no temporary functions. I've since learned that the special `pg_temp` schema always refers to the current session's temporary objects, so creating a function there ensures it cannot collide with anything else and will not live past session end.

In actuality, each session has a numbered temporary schema (e.g. `pg_temp_2`) and `pg_temp` is just an alias for the correct one for your session.